features/nanostack/FEATURE_NANOSTACK/mbed-mesh-api/source/ThreadInterface.cpp

```diff
@@ -43,10 +43,20 @@ int ThreadInterface::connect()
     // -routers will create new network and get local connectivity
     // -end devices will get connectivity once attached to existing network
     // -devices without network settings gets connectivity once commissioned and attached to network
-    int32_t count = connect_semaphore.wait(osWaitForever);
-
-    if (count <= 0) {
-        return NSAPI_ERROR_DHCP_FAILURE; // sort of...
+    _connect_status = NSAPI_STATUS_CONNECTING;
+    if (_connection_status_cb) {
+        _connection_status_cb(NSAPI_EVENT_CONNECTION_STATUS_CHANGE, NSAPI_STATUS_CONNECTING);
+    }
+    if (_blocking) {
+        int32_t count = connect_semaphore.wait(osWaitForever);
+
+        if (count <= 0) {
```

Comment:
As it's a "wait forever", not sure if this failure case is really required. In any case, if we give up waiting, that doesn't mean we're not still connecting, so the state change isn't correct, I think. The state should only change from the callbacks.